### PR TITLE
Branch graphs.

### PR DESCRIPTION
**Pull request by tinevez**
# Branch graph: a branch simplification view of a source graph. 

This class implements a view of a source `ListenableGraph` (called here **linked graph**) that offers a coarser level of details by representing branches of the linked graph as a single vertex and edge in the branch graph. 

A branch in the linked graph is defined as connected vertices that have exactly 1 incoming edge and 1 outgoing edge. In the branch graph, a branch is represented as follow. If in the linked graph a branch is like this: 
```
 v0 → v1 → v2 → v3
 ```
The its representation in the branch graph is:   `bv0 → bv1`
 
where: 

- `bv0` links to `v0`, 
- `bv1` links to `v3`, 
- all linked vertices in-between linked to the branch edge between `bv0` and `bv1`, 
- as well as all the linked edges in the branch. 
- The branch edge between `bv0` and `bv1` links to the outgoing edge of `v0`. 

For instance, a graph laid as following: 
```
                v3 → v4 → v5
              /
 v0 → v1 → v2
              \
                 v6 → v7 → v8
 ```

will be represented by the following branch graph:
```
            bv2
          /
 bv0 → bv1
          \
            bv3
```
 
In the example above, `v0`, `v2`, `v5` and `v8` are branch extremities; they are linked to a unique branch vertex in the branch graph. `v1`, `v3`, `v4`, `v6` and `v7` belongs to a branch. They link to a branch edge. 

The branch graph is based on a non-simple directed graph. There might be more than one edge between the same source and target branch vertices. This happens for instance when there is a diamond-like shape in the linked graph: 
```
    v1 → v2 → v3
   /             \
 v0               v7
   \             /
     v4 → v5 → v6
 ```

In the branch graph, this becomes:
```
     ____
    /    \
 bv0      bv1
    \____/
```
 
The branch graph can also handle ring-link structures in the linked graph. In that case, such a ring;
```
     v1 → v2 → v3
   /             \
 v0               v4
   \             /
     v7 ← v6 ← v5
```
 
is represented by a single vertex having a loop-edge:
```
    _
 bv0 \
  \__/
```
 
In such a loop, the vertex linked to `bv0` and the edge linked to the branch edge are determined by the order in which the vertices and edges are added in the linked graph. 
The branch graph is defined and tested only for simple, directed graphs as linked graph. Using any other classes of graphs will result in unexpected behavior.

### CODE REVIEW

**Comment by tpietzsch** [comment]
@tinevez I'm reviewing it now. Here is some changes I made already:
* Make AbstractEdgePool class abstract (again)
* Move AbstractEdgeLayout to AbstractEdge (from AbstractSimpleEdge)
* Remove AbstractSimpleEdge class. It adds nothing over AbstractEdge and is not required

Request that I have already:

1.)
Revise naming of `src/test/java/org/mastodon/graph/TestNonSimpleEdge.java` etc and look through the code of those again. `TestNonSimpleEdge` should be just `TestSimpleEdge`. `TestEdgePool` should be `TestSimpleEdgePool`.

For many of these, they might not even be required? The difference between simple and non-simple graphs is really only in the edge pool. Vertices, edges, etc should not need to be different between the two. A edge doesn't need to know whether it is simple or non-simple. I didn't look through all the test classes in detail. Can you just have another look at everything there?

2.)
As it is `ReadOnlyGraph` (which now also applies to non-simple graphs) has no way to get at all the edges between two vertices. There is just `getEdge(source, target)`, which returns the first such edge. Consider whether it makes sense to add `getEdges(source, target)` and what this should return to minimize new object instantiation (an iterator over the edges, reusable iterator specified as `ref` parameter somehow)? I have no clear answer so far. Maybe just think about it and then let's discus...

File: src/main/java/org/mastodon/graph/branch/BranchGraph.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	BV extends Vertex< BE >,
+	BE extends Edge< BV >,
+	V extends Vertex< E >,
+	E extends Edge< V > >
+		extends ListenableReadOnlyGraph< BV, BE >
+{
+
+	/**
+	 * Returns the edge linked to the specified branch edge. The linked edge is
+	 * the single outgoing edge of the branch extremity.
```
Comment (tpietzsch) [inline src/main/java/org/mastodon/graph/branch/BranchGraph.java]: It is not necessarily true, that a branch extremity has a single outgoing edge. (?) Can you make this more clear?

File: src/main/java/org/mastodon/graph/branch/BranchGraph.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	 * @param ref
+	 *            a reference object to the branch edge used for retrieval.
+	 * @return a branch edge.
+	 */
+	public BE getBranchEdge( E edge, BE ref );
+
+	/**
+	 * Returns the branch vertex linked to the specified edge if it belongs to a
+	 * branch. Returns <code>null</code> if the specified vertex is a branch
+	 * extremity.
```
Comment (tpietzsch) [inline src/main/java/org/mastodon/graph/branch/BranchGraph.java]: This seems to confuse edge/vertex. "Returns the branch vertex", but returns a branch edge. "...to the specified edge", but takes a vertex argument. Please revise.

File: src/main/java/org/mastodon/graph/branch/BranchGraph.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	 * Returns a graph id map for the branch graph.
+	 * 
+	 * @return a graph id map.
+	 */
+	public GraphIdBimap< BV, BE > getGraphIdBimap();
+
+	/**
+	 * Returns an iterator that iterates in order over the linked vertices in a
+	 * branch, specified by its branch edge. The first and last vertex iterated
+	 * are the branch nodes.
```
Comment (tpietzsch) [inline src/main/java/org/mastodon/graph/branch/BranchGraph.java]: "...are the branch nodes" should be "branch extremities"?

**Comment by tinevez** [comment]
@tpietzsch Here are first elements of answer.

Regarding 1.:
I changed the name of the test classes as well. This was a big oversight on my part. 
I looked and I think the `TestSimpleEdge` and `TestEdge` distinction is needed. A concrete edge knows about its concrete test edge pool (in the constructor and in the generic types), so you have to create a specific edge class for a specific graph.

Regarding 2.:
For the return type of `getEdges(source, target)`, a `Collection` might be better than an iterator since with it the user can immediately know the number of edges in between two vertices. I will come up with a tentative. 
Is it ok to do on this PR?

**Comment by tpietzsch** [comment]
> For the return type of getEdges(source, target), a Collection might be better than an iterator since with it the user can immediately know the number of edges in between two vertices. I will come up with a tentative.

Sounds good!

> Is it ok to do on this PR?

Yes, please do.

**Comment by tinevez** [comment]
I added `getEdges()` method to the `ReadOnlyGraph` interface, with implementations for `GraphImp` and `AbstractGraphObject`. Only the first one is under JUnit tests yet.

They return an iterator, because it was simpler to start with ._.
For `GraphImp`, the iterator reference can be reused in method that accept a `ref` iterator, provided that it was created by the `getEdges()` method first. It minimises instantiations, building on the `OutGoingEdges` class.

What do you think?

**Comment by tpietzsch** [comment]
> They return an iterator, because it was simpler to start with

I modified it to return `Edges<E>` which seems the canonical way to do it. It implemented and tested for `GraphImp`.

Could you:
    * implement it for object graphs, and
    * add javadoc for `getEdges()` in `ReadOnlyGraph` interface?

**Comment by tpietzsch** [comment]
Awesome, thanks @tinevez !

**Pull request merged**